In [67]:
import pandas as pd
import numpy as np
from scipy import stats
from itertools import combinations

# Hypothesis Testing Exercise

## Q1     
     A F&B manager wants to determine whether there is any significant difference in the diameter of the cutlet between two units. A randomly selected sample of cutlets was collected from both units and measured? Analyze the data and draw inferences at 5% significance level. Please state the assumptions and tests that you carried out to check validity of the assumptions.


Minitab File : Cutlets.mtw


In [19]:
df = pd.read_csv("D:\\Samit\\2023\\DS_ExcelR\\Assignments\\Hypothesis Testing\\Cutlets.csv")
df.head()

Unit A  Unit B
0  6.8090  6.7703
1  6.4376  7.5093
2  6.9157  6.7300
3  7.3012  6.7878
4  7.4488  7.1522

### Assumptions
    Ho: There is no significant difference in the diameter of the cutlet between two units UnitA and UnitB
    H1: There is significant difference in the diameter of the cutlet between two units UnitA and UnitB
    Significance level = 0.05

In [20]:
stats.ttest_ind(df['Unit A'],df['Unit B'])                 #Here ttest_ind() is the function two calculate statistics and p-value for two variables, ind represents independent

Ttest_indResult(statistic=0.7228688704678063, pvalue=0.47223947245995)

In [22]:
print("p-value: ",stats.ttest_ind(df['Unit A'],df['Unit B'])[1])

p-value:  0.47223947245995


### Conclusion
    Here, Since p-value: 0.4722 > significance level: 0.05.
    So, the conclusion is Null hypothesis is accepted which means there is no significant difference in the diameter of the cutlet between two units UnitA and UnitB

## Q2
    A hospital wants to determine whether there is any difference in the average Turn Around Time (TAT) of reports of the laboratories on their preferred list. They collected a random sample and recorded TAT for reports of 4 laboratories. TAT is defined as sample collected to report dispatch.
   
    Analyze the data and determine whether there is any difference in average TAT among the different laboratories at 5% significance level.
 
 
    Minitab File: LabTAT.mtw


In [68]:
df = pd.read_csv('D:\\Samit\\2023\\DS_ExcelR\\Assignments\\Hypothesis Testing\\LabTAT.csv')
df

Laboratory 1  Laboratory 2  Laboratory 3  Laboratory 4
0          185.35        165.53        176.70        166.13
1          170.49        185.91        198.45        160.79
2          192.77        194.92        201.23        185.18
3          177.33        183.00        199.61        176.42
4          193.41        169.57        204.63        152.60
..            ...           ...           ...           ...
115        178.49        170.66        193.80        172.68
116        176.08        183.98        215.25        177.64
117        202.48        174.54        203.99        170.27
118        182.40        197.18        194.52        150.87
119        182.09        215.17        221.49        162.21

[120 rows x 4 columns]

### Assumptions
    Ho: There is no difference in the average Turn Around Time (TAT) of reports of the laboratories on their preferred list.
    H1: There is difference in the average Turn Around Time (TAT) of reports of the laboratories on their preferred list.

In [69]:
stats.f_oneway(df.iloc[:,0],df.iloc[:,1],df.iloc[:,2],df.iloc[:,3])

F_onewayResult(statistic=118.70421654401437, pvalue=2.1156708949992414e-57)

In [70]:
print("p-value: ",stats.f_oneway(df.iloc[:,0],df.iloc[:,1],df.iloc[:,2],df.iloc[:,3])[1])

p-value:  2.1156708949992414e-57


### Conclusion
    Here since p-value: 2.1156708949992414e-57 < significance level: 0.05, so alternate hypothesis is accepted. Which means that there is difference in the average Turn Around Time (TAT) of reports of the laboratories on their preferred list. 

## Q3
    Sales of products in four different regions is tabulated for males and females. Find if male-female buyer rations are similar across regions.

    Full question is in PPTX.


### Assumptions

    Ho: Male and Female proportions are equal across all regions
    H1: Not all Male and Female proportions are equal across all regions

    Assume alpha = 0.05 

In [71]:
df = pd.read_csv("D:\\Samit\\2023\\DS_ExcelR\\Assignments\\Hypothesis Testing\\BuyerRatio.csv")
df

Observed Values  East  West  North  South
0           Males    50   142    131     70
1         Females   435  1523   1356    750

In [72]:
df.set_index(df['Observed Values'],inplace=True)
df

Observed Values  East  West  North  South
Observed Values                                          
Males                     Males    50   142    131     70
Females                 Females   435  1523   1356    750

In [73]:
del df['Observed Values']

In [74]:
df

East  West  North  South
Observed Values                          
Males              50   142    131     70
Females           435  1523   1356    750

In [75]:
observed_data = df.to_numpy()
observed_data

array([[  50,  142,  131,   70],
       [ 435, 1523, 1356,  750]], dtype=int64)

In [76]:
stats.chi2_contingency(observed_data)            # chi2_contingency() function for chi-square test

Chi2ContingencyResult(statistic=1.595945538661058, pvalue=0.6603094907091882, dof=3, expected_freq=array([[  42.76531299,  146.81287862,  131.11756787,   72.30424052],
       [ 442.23468701, 1518.18712138, 1355.88243213,  747.69575948]]))

In [77]:
chi2,p,dof,expected = stats.chi2_contingency(observed_data)
print("Chi-Square statistics: ",chi2)
print("p-value: ",p)
print("Degrees of freedom: ",dof)
print("Expected frequencies: ")
print(expected)

Chi-Square statistics:  1.595945538661058
p-value:  0.6603094907091882
Degrees of freedom:  3
Expected frequencies: 
[[  42.76531299  146.81287862  131.11756787   72.30424052]
 [ 442.23468701 1518.18712138 1355.88243213  747.69575948]]


### Conclusion
    Here, p-value: 0.66 > significance level 0.05, so we accept null hypothesis which means that Male and Female proportions are equal across all regions. 

## Q4
    TeleCall uses 4 centers around the globe to process customer order forms. They audit a certain %  of the customer order forms. Any error in order form renders it defective and has to be reworked before processing.  The manager wants to check whether the defective %  varies by centre. Please analyze the data at 5% significance level and help the manager draw appropriate inferences

Minitab File: CustomerOrderForm.mtw


In [78]:
df = pd.read_csv("D:\\Samit\\2023\\DS_ExcelR\\Assignments\\Hypothesis Testing\\Costomer+OrderForm.csv")
df

Phillippines   Indonesia       Malta       India
0     Error Free  Error Free   Defective  Error Free
1     Error Free  Error Free  Error Free   Defective
2     Error Free   Defective   Defective  Error Free
3     Error Free  Error Free  Error Free  Error Free
4     Error Free  Error Free   Defective  Error Free
..           ...         ...         ...         ...
295   Error Free  Error Free  Error Free  Error Free
296   Error Free  Error Free  Error Free  Error Free
297   Error Free  Error Free   Defective  Error Free
298   Error Free  Error Free  Error Free  Error Free
299   Error Free   Defective   Defective  Error Free

[300 rows x 4 columns]

### Assumptions
    Ho: The defective % does not varies by centre.
    H1: The defective % varies by centre. 

In [79]:
column_combinations = list(combinations(df.columns,2))
column_combinations

[('Phillippines', 'Indonesia'),
 ('Phillippines', 'Malta'),
 ('Phillippines', 'India'),
 ('Indonesia', 'Malta'),
 ('Indonesia', 'India'),
 ('Malta', 'India')]

In [80]:
for comb in column_combinations:
    contingency_table = pd.crosstab(df[comb[0]],df[comb[1]])
    # print(contingency_table,"\n")
    chi2, p, dof, expected = stats.chi2_contingency(contingency_table)

    print(f"\nChi-square test for columns: {comb}")
    print("Chi-square statistics: ", chi2)
    print("p-value: ", p)
    print("Degrees of freedom: ", dof)
    print("Expected frequencies: ",)
    print(expected)


Chi-square test for columns: ('Phillippines', 'Indonesia')
Chi-square statistics:  0.1856391005881107
p-value:  0.6665712150680798
Degrees of freedom:  1
Expected frequencies: 
[[  3.19  25.81]
 [ 29.81 241.19]]

Chi-square test for columns: ('Phillippines', 'Malta')
Chi-square statistics:  0.1043743747062788
p-value:  0.7466420539639729
Degrees of freedom:  1
Expected frequencies: 
[[  2.99666667  26.00333333]
 [ 28.00333333 242.99666667]]

Chi-square test for columns: ('Phillippines', 'India')
Chi-square statistics:  0.0
p-value:  1.0
Degrees of freedom:  1
Expected frequencies: 
[[  1.93333333  27.06666667]
 [ 18.06666667 252.93333333]]

Chi-square test for columns: ('Indonesia', 'Malta')
Chi-square statistics:  9.52050868943337
p-value:  0.0020318825161213722
Degrees of freedom:  1
Expected frequencies: 
[[  3.41  29.59]
 [ 27.59 239.41]]

Chi-square test for columns: ('Indonesia', 'India')
Chi-square statistics:  0.0
p-value:  1.0
Degrees of freedom:  1
Expected frequencies: 
[[ 

    Since, there are six different p-values for each combination. So to draw conclusions from them we will be using Bonferroni correction 

    Below is shown how to apply Bonferroni correction

In [81]:
alpha = 0.05
alpha_adj = alpha/6    #6 p-value are there, so dividing alpha into 6 part
alpha_adj

0.008333333333333333

### Evaluation
    So now comparing above alpha_adj(~ 0.0083) value with all combinations p-values 

    So, for column: ('Phillippines', 'Indonesia'):
    p-value: 0.666 > alpha_adj: 0.0083
    So, we accept null hypothesis which means defective % does not varies by centre.

    for column: ('Phillippines', 'Malta')
    p-value: 0.746 > alpha_adj: 0.0083
    So, we accept null hypothesis which means defective % does not varies by centre.

    for column: ('Phillippines', 'India')
    p-value: 1 > alpha_adj: 0.0083
    So, we accept null hypothesis which means defective % does not varies by centre.

    for column: ('Indonesia', 'Malta')
    p-value: 0.0020 < alpha_adj: 0.0083
    So, we accept alternate hypothesis which means defective % varies by centre.

    for column: ('Indonesia', 'India')
    p-value: 1 > alpha_adj: 0.0083
    So, we accept null hypothesis which means defective % does not varies by centre.

    for column: ('Malta', 'India')   
    p-value: 0.233 > alpha_adj: 0.0083
    So, we accept null hypothesis which means defective % does not varies by centre.
     

### Conclusion
    So column: ('Indonesia', 'Malta')
    p-value: 0.0020 < alpha_adj: 0.0083
    is the combination where alternate hypothesis is accepted.


    So, overall conclusion would be there is evidence of an association in one combination between Indonesia and Malta which means defective % varies by centre, but no evidence of association in the other combinations which means defective % does not varies by centre.

    By Samit Dhawal